src/api/spec/models/unregistered_user_spec.rb

```diff
@@ -0,0 +1,124 @@
+require 'rails_helper'
+
+RSpec.describe UnregisteredUser do
+  let(:admin_user) { create(:admin_user, login: 'king') }
+  let(:user) { create(:user, login: 'eisendieter') }
+  let(:confirmed_user) { create(:confirmed_user, login: 'confirmed_user') }
+  let(:input) { { 'Event::RequestCreate' => { source_maintainer: '1' } } }
+  let(:project_with_package) { create(:project_with_package, name: 'project_b') }
+
+  describe 'validations' do
+    it { is_expected.to validate_presence_of(:login).with_message('must be given') }
+    it { is_expected.to validate_length_of(:login).is_at_least(2).with_message('must have more than two characters') }
+    it { is_expected.to validate_length_of(:login).is_at_most(100).with_message('must have less than 100 characters') }
+    it { is_expected.to validate_inclusion_of(:state).in_array(User::STATES) }
+
+    it { is_expected.to allow_value('king@opensuse.org').for(:email) }
+    it { is_expected.to_not allow_values('king.opensuse.org', 'opensuse.org', 'opensuse').for(:email) }
+
+    it { expect(user.state).to eq('unconfirmed') }
+
+    it { expect(create(:user)).to validate_uniqueness_of(:login).with_message('is the name of an already existing user') }
+  end
+
+  context '#register' do
+    let(:user_attributes) do
+      {
+        realname:              'Tux Penguin',
+        login:                 'tux',
+        password:              'tux123',
+        password_confirmation: 'tux123',
+        email:                 'tux@northpole.org'
+      }
+    end
+    let(:attributes_for_query) { user_attributes.slice(:login, :realname, :email).merge(state: 'confirmed') }
+
+    context 'when not in LDAP mode' do
+      context 'when registration is allowed' do
+        before do
+          allow(::Configuration).to receive(:registration).and_return('allow')
+        end
+
+        subject! { UnregisteredUser.register(user_attributes) }
+
+        it 'creates a new confirmed user' do
+          expect(User.where(attributes_for_query)).to exist
+        end
+      end
+
+      context 'when registration requires confirmation' do
+        let(:attributes_for_query) { user_attributes.slice(:login, :realname, :email).merge(state: 'unconfirmed') }
+
+        before do
+          allow(::Configuration).to receive(:registration).and_return('confirmation')
+        end
+
+        subject { UnregisteredUser.register(user_attributes) }
+
+        it 'throws an exception that confirms the user registration... and creates an unconfirmed user' do
+          expect { subject }.to raise_error(UnregisteredUser::ErrRegisterSave,
+                                            'Thank you for signing up! An admin has to confirm your account now. Please be patient.')
+          expect(User.where(attributes_for_query)).to exist
```

Comment:
Why does the user exists? I would expect that the user does not exist or at least is unconfirmed?

Reply:
He exists, but is unconfirmed. That's checked via the attributes_for_query (see line 51)

Reply:
Ok, missed the update of the attributes for query variable